x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+) *Keeper {
+
+	k := Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		tStoreKey:     tStoreKey,
+		authority:     authority,
+		spKeeper:      spKeeper,
+		accountKeeper: accountKeeper,
+		bankKeeper:    bankKeeper,
+		paymentKeeper: paymentKeeper,
+	}
+
+	k.lvgSequence = sequence.NewSequence[uint32](types.LVGSequencePrefix)
+	k.gvgSequence = sequence.NewSequence[uint32](types.GVGSequencePrefix)
+	k.gvgFamilySequence = sequence.NewSequence[uint32](types.GVGFamilySequencePrefix)
+
+	return &k
+}
+
+func (k Keeper) GetAuthority() string {
+	return k.authority
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) GenNextLVGID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.lvgSequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) GenNextGVGID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.gvgSequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) GenNextGVGFamilyID(ctx sdk.Context) uint32 {
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.gvgFamilySequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) SetGVG(ctx sdk.Context, gvg *types.GlobalVirtualGroup) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvg)
+	store.Set(types.GetGVGKey(gvg.Id), bz)
+}
+
+func (k Keeper) DeleteGVG(ctx sdk.Context, primarySpID, gvgID uint32) error {
+
+	store := ctx.KVStore(k.storeKey)
+
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return types.ErrGVGNotExist
+	}
+
+	if gvg.StoredSize != 0 {
+		return types.ErrGVGNotEmpty
+	}
+
+	gvgFamily, found := k.GetGVGFamily(ctx, primarySpID, gvg.FamilyId)
+	if !found {
+		panic("not found gvg family when delete gvg")
+	}
+
+	err := gvgFamily.RemoveGVG(gvg.Id)
+	if err == types.ErrGVGNotExist {
+		panic("gvg not found in gvg family when delete gvg")
+	}
+
+	for _, secondarySPID := range gvg.SecondarySpIds {
+		gvgStatisticsWithinSP := k.MustGetGVGStatisticsWithinSP(ctx, secondarySPID)
+		gvgStatisticsWithinSP.SecondaryCount--
+		k.SetGVGStatisticsWithSP(ctx, gvgStatisticsWithinSP)
+	}
+
+	store.Delete(types.GetGVGKey(gvg.Id))
+
+	k.SetGVGFamily(ctx, gvg.PrimarySpId, gvgFamily)
+	return nil
+}
+
+func (k Keeper) GetGVG(ctx sdk.Context, gvgID uint32) (*types.GlobalVirtualGroup, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGKey(gvgID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var gvg types.GlobalVirtualGroup
+	k.cdc.MustUnmarshal(bz, &gvg)
+	return &gvg, true
+}
+
+func (k Keeper) GetGVGByLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32) (*types.GlobalVirtualGroup, bool) {
+	lvg, found := k.GetLVG(ctx, bucketID, lvgID)
+	if !found {
+		return nil, false
+	}
+	gvg, found := k.GetGVG(ctx, lvg.GlobalVirtualGroupId)
+	if !found {
+		return nil, false
+	}
+	return gvg, true
+}
+
+func (k Keeper) SetLVG(ctx sdk.Context, lvg *types.LocalVirtualGroup) {
+	store := ctx.KVStore(k.storeKey)
```

Comment:
As we expected there to be many `lvg`, so we'd better save the sapce here:
```
lvg.Id = 0
lvg.BucketId = math.ZeroUint()
```
And ensure the `lvg` never be used after `SetLVG`. (as the id and bucket id is set to zero)

Reply:
as we use `bucketID` as part of the key, whose type is Uint256, any method that we can compress the key?  (it seems long and cause a fat node in iavl store)

Reply:
The `Uint256` stores not as a fixed bytes.

```
// Bytes returns the absolute value of x as a big-endian byte slice.
//
// To use a fixed length slice, or a preallocated one, use FillBytes.
func (x *Int) Bytes() []byte {
	// This function is used in cryptographic operations. It must not leak
	// anything but the Int's sign and bit size through side-channels. Any
	// changes must be reviewed by a security expert.
	buf := make([]byte, len(x.abs)*_S)
	return buf[x.abs.bytes(buf):]
}
```

Reply:
> As we expected there to be many `lvg`, so we'd better save the sapce here:
> 
> ```
> lvg.Id = 0
> lvg.BucketId = math.ZeroUint()
> ```
> 
> And ensure the `lvg` never be used after `SetLVG`. (as the id and bucket id is set to zero)


Nice. We can do this optimization and decode `lvgid` and `bucketid` from the key when we get it